op-node/sources/fallback_client.go

```diff
@@ -0,0 +1,243 @@
+	fallbackClient := &FallbackClient{
+		ctx:          ctx,
+		firstRpc:     rpc,
+		urlList:      urlList,
+		log:          log,
+		rpcInitFunc:  rpcInitFunc,
+		currentRpc:   rpc,
+		currentIndex: 0,
+		l1ChainId:    l1ChainId,
+		l1Block:      l1Block,
+	}
+	go func() {
+		ticker := time.NewTicker(1 * time.Minute)
+		for {
+			select {
+			case <-ticker.C:
+				log.Debug("FallbackClient clear lastMinuteFail 0")
+				fallbackClient.lastMinuteFail.Store(0)
+			}
+		}
+	}()
+	return fallbackClient
+}
+
+func (l *FallbackClient) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	return l.EthSubscribe(ctx, ch, "newHeads")
+}
+
+func (l *FallbackClient) Close() {
+	l.currentRpc.Close()
+	if l.currentRpc != l.firstRpc {
+		l.firstRpc.Close()
+	}
+}
+
+func (l *FallbackClient) CallContext(ctx context.Context, result any, method string, args ...any) error {
+	err := l.currentRpc.CallContext(ctx, result, method, args...)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return err
+}
+
+func (l *FallbackClient) handleErr(err error) {
+	if err == rpc.ErrNoResult {
+		return
+	}
+	errCount := l.lastMinuteFail.Add(1)
+	if errCount > 10 {
+		l.switchCurrentRpc()
+	}
+}
+
+func (l *FallbackClient) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
+	err := l.currentRpc.BatchCallContext(ctx, b)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return err
+}
+
+func (l *FallbackClient) EthSubscribe(ctx context.Context, channel any, args ...any) (ethereum.Subscription, error) {
+	subscribe, err := l.currentRpc.EthSubscribe(ctx, channel, args...)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return subscribe, err
+}
+
+func (l *FallbackClient) switchCurrentRpc() {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+	if l.lastMinuteFail.Load() <= 10 {
+		return
+	}
+	for {
+		l.currentIndex++
+		err := l.switchCurrentRpcLogic()
+		if err != nil {
+			l.log.Warn("the fallback client failed to switch the current client", "err", err)
+		} else {
+			break
+		}
+	}
+}
+
+func (l *FallbackClient) switchCurrentRpcLogic() error {
+	if l.currentIndex >= len(l.urlList) {
+		return fmt.Errorf("the fallback client has tried all urls")
+	}
+	url := l.urlList[l.currentIndex]
+	newRpc, err := l.rpcInitFunc(url)
+	if err != nil {
+		return fmt.Errorf("the fallback client init RPC failed,url:%s, err:%v", url, err)
+	}
+	vErr := l.validateRpc(newRpc)
+	if vErr != nil {
+		return vErr
+	}
+	lastRpc := l.currentRpc
+	l.currentRpc = newRpc
+	if lastRpc != l.firstRpc {
+		lastRpc.Close()
+	}
+	l.lastMinuteFail.Store(0)
+	if l.subscribeFunc != nil {
+		l.reSubscribeNewRpc(url)
+	}
+	l.log.Info("switched current rpc to new url", "url", url)
+	if !l.isInFallbackState {
+		l.isInFallbackState = true
+		l.recoverIfFirstRpcHealth()
+	}
+	return nil
+}
+
+func (l *FallbackClient) reSubscribeNewRpc(url string) {
```

Comment:
Can you confirm that this function works? For example, if we subscribe new heads using the subscription feature, after switching the RPC client and invoking `reSubscribeNewRpc` here, will we still get new heads? Will there be any missing or duplicate heads in the results? If so, are the cases handled properly?

Reply:
Yes, I have tested this scenario. When the url is switched, you can see that the l1 head is growing normally through the json_rpc call of optimism_syncStatus.   
However, since we need to check the health of the first rpc, after Unsubscribe, we can still observe that the first url is requested periodically and error logs are printed. This is because PollingClient's polling will not be stopped by Unsubscribe. Anyway, it doesn't affect anything except printing error logs.